pyeudiw/sd_jwt/holder.py

```diff
@@ -0,0 +1,255 @@
+                            f"Check disclosure information for array: {sd_jwt_claims}"
+                        )
+
+                    self._select_disclosures(
+                        disclosure_value, claims_to_disclose_element
+                    )
+
+            else:
+                self._select_disclosures(element, claims_to_disclose_element)
+
+    def _select_disclosures_dict(self, sd_jwt_claims, claims_to_disclose):
+        if claims_to_disclose is None:
+            return {}
+        if claims_to_disclose is True:
+            # Tolerate a "True" for a disclosure of an object
+            claims_to_disclose = {}
+        if not isinstance(claims_to_disclose, dict):
+            raise ValueError(
+                f"To disclose object elements, an object must be provided as disclosure information.\n"
+                f"Found {claims_to_disclose} (type {type(claims_to_disclose)}) instead.\n"
+                f"Check disclosure information for object: {sd_jwt_claims}"
+            )
+        for key, value in sd_jwt_claims.items():
+            if key == SD_DIGESTS_KEY:
+                for digest_to_check in value:
+                    if digest_to_check not in self._hash_to_decoded_disclosure:
+                        # fake digest
+                        continue
+                    _, key, value = self._hash_to_decoded_disclosure[digest_to_check]
+
+                    try:
+                        logger.debug(
+                            f"In _select_disclosures_dict: {key}, {value}, {claims_to_disclose}"
+                        )
+                        if key in claims_to_disclose and claims_to_disclose[key]:
+                            logger.debug(f"Adding disclosure for {digest_to_check}")
+                            self.hs_disclosures.append(
+                                self._hash_to_disclosure[digest_to_check]
+                            )
+                        else:
+                            logger.debug(
+                                f"Not adding disclosure for {digest_to_check}, {key} (type {type(key)}) not in {claims_to_disclose}"
+                            )
+                    except TypeError:
+                        # claims_to_disclose is not a dict
+                        raise TypeError(
+                            f"claims_to_disclose does not contain a dict where a dict was expected (found {claims_to_disclose} instead)\n"
+                            f"Check claims_to_disclose for key: {key}, value: {value}"
+                        ) from None
+
+                    self._select_disclosures(value, claims_to_disclose.get(key, None))
+            else:
+                self._select_disclosures(value, claims_to_disclose.get(key, None))
+
+    def _create_key_binding_jwt(
+        self, nonce, aud, presentation_hash, holder_key, sign_alg: Optional[str] = None
+    ):
+        _alg = sign_alg or DEFAULT_SIGNING_ALG
+
+        self.key_binding_jwt_header = {
```

Comment:
Here there is a small bug.

While the holder key binding JWT is correct, the signer downstream will add a claim `kid` to the JWT header based on the signing key kid.
Key binding JWT do not require a kid header, see the [reference spec](https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-14.html#name-key-binding-jwt). This is because there can only ever be one `cnf` key in a given issuer token and, as such, it can always be identified without ambiguity; see the relevant section in [RFC7800](https://www.rfc-editor.org/rfc/rfc7800.html#section-3.1). Likewise, the confirmation key in expressed in the `cnf` claim of the issuer token is written without kid.

Also, see https://github.com/italia/eudi-wallet-it-docs/issues/411

Reply:
true

when jwk is present in the header, kid MUST NOT be added

the same when x5c is present

unless a general configuration parameter might override this default behaviour, as discussed to @LadyCodesItBetter